vortex-layout/src/layouts/stats/reader.rs

```diff
@@ -1,167 +1,187 @@
 use std::ops::Range;
-use std::sync::{Arc, RwLock};
+use std::sync::{Arc, OnceLock, RwLock};
 
-use async_once_cell::OnceCell;
+use futures::future::{BoxFuture, Shared};
+use futures::{FutureExt, TryFutureExt};
 use vortex_array::ArrayContext;
-use vortex_array::aliases::hash_map::HashMap;
+use vortex_array::aliases::hash_map::{Entry, HashMap};
 use vortex_array::stats::{Stat, stats_from_bitset_bytes};
 use vortex_dtype::TryFromBytes;
-use vortex_error::{VortexExpect, VortexResult, vortex_bail, vortex_err, vortex_panic};
+use vortex_error::{SharedVortexResult, VortexExpect, VortexResult, vortex_panic};
 use vortex_expr::pruning::PruningPredicate;
 use vortex_expr::{ExprRef, Identity};
 use vortex_mask::Mask;
 
 use crate::layouts::stats::StatsLayout;
 use crate::layouts::stats::stats_table::StatsTable;
 use crate::reader::LayoutReader;
-use crate::segments::AsyncSegmentReader;
-use crate::{ExprEvaluator, Layout, LayoutVTable, RowMask};
+use crate::segments::SegmentSource;
+use crate::{ExprEvaluator, Layout, LayoutVTable};
 
-type PruningCache = Arc<OnceCell<Option<Mask>>>;
+pub(crate) type SharedStatsTable = Shared<BoxFuture<'static, SharedVortexResult<StatsTable>>>;
+pub(crate) type SharedPruningResult = Shared<BoxFuture<'static, SharedVortexResult<Option<Mask>>>>;
+pub(crate) type PredicateCache = Arc<OnceLock<Option<PruningPredicate>>>;
 
-#[derive(Clone)]
 pub struct StatsReader {
     layout: Layout,
-    ctx: ArrayContext,
-    segment_reader: Arc<dyn AsyncSegmentReader>,
-
-    /// The number of blocks
-    nblocks: usize,
-    /// The size of each block (except possibly the last)
-    block_len: usize,
-    /// The stats present in the layout
-    present_stats: Arc<[Stat]>,
+
+    /// Data layout reader
+    pub(crate) data_child: Arc<dyn LayoutReader>,
+    /// Stats table layout reader.
+    pub(crate) stats_child: Arc<dyn LayoutReader>,
+
+    /// The number of zones
+    pub(crate) nzones: usize,
+    /// The size of each zone (except possibly the last)
+    pub(crate) zone_len: usize,
```

Comment:
nit - somewhat incosistent naming? comment should probably be something like "The number of rows in each zone (except possibly the last)"